(maint) Add contributing and code of conduct docs (#7)
puppetlabs/puppetlabs-http_request: CONTRIBUTING.md

# Contributing to Puppet modules

So you want to contribute to a Puppet module: Great! Below are some instructions

> to get you started doing that very thing while setting expectations around code

Reviewer: ```suggestion
to get you started while setting expectations around code
```

Reviewer: Actually, it'd be good to have Heston review this for stuff like this.

> So you want to contribute to a Puppet module: Great! Below are some instructions
> to get you started while setting expectations around code quality, as well as a
> few tips for making the process as easy as possible. 

Reviewer 2: ```suggestion
So you want to contribute to a Puppet module? Great! Follow our guidelines
to familiarize yourself with our expectations around code quality, and learn some tips
to make the contribution process as easy as possible.
```

> quality as well as a few tips for making the process as easy as possible. 

Reviewer: ```suggestion
quality, as well as a few tips for making the process as easy as possible. 
```

> ## Community Slack Channels

Reviewer 2: ```suggestion
## Community Slack channels
```

Join the `#bolt` channel in the [Puppet community
Slack](https://slack.puppet.com/) where Bolt developers and community members
who use and contribute to Bolt discuss the tool.

## Getting started

> - Fork the module repository on GitHub and clone to your workspace

Reviewer 2: ```suggestion
- Fork the module repository on GitHub and clone to your workspace.
```

- Make your changes!

## Commit checklist

### The basics

> - My commit is a single logical unit of work
> - I have checked for unnecessary whitespace with `git diff --check`
> - My commit does not include commented out code or unneeded files

Reviewer 2: ```suggestion
- My commit is a single logical unit of work.
- I have checked for unnecessary whitespace with `git diff --check`.
- My commit does not include commented out code or unneeded files.
```

### The content

> - My commit includes tests for the bug I fixed or feature I added
> - My commit includes appropriate documentation changes if it is introducing a
>   new feature or changing existing functionality
> - My code passes existing test suites

Reviewer 2: ```suggestion
- My commit includes tests for the bug I fixed or feature I added.
- My commit includes appropriate documentation changes if it is introducing a
  new feature or changing existing functionality.
- My code passes existing test suites.
```

### The commit message

> - The first line of my commit message includes:
>   - An issue number (if applicable), e.g. "(GH-xxxx) This is the first line"
>   - A short description (50 characters is the soft limit, excluding ticket
>     number(s))
> - The body of my commit message:
>   - Is meaningful
>   - Uses the imperative, present tense: "change", not "changed" or "changes"
>   - Includes motivation for the change, and contrasts its implementation with
>     the previous behavior

Reviewer 2: ```suggestion
- The first line of my commit message includes:
  - An issue number (if applicable). For example, `(GH-xxxx) This is the first line`.
  - A short description (50 characters is the soft limit, excluding ticket
    number(s)).
- The body of my commit message:
  - Is meaningful.
  - Uses the imperative, present tense: "change", not "changed" or "changes".
  - Includes motivation for the change, and contrasts its implementation with
    the previous behavior.
```

## Testing

### Getting started

Our Puppet modules provide [`Gemfile`](./Gemfile)s, which can tell a Ruby
package manager such as [bundler](http://bundler.io/) what Ruby packages, or

> Gems, are required to build, develop, and test this software.

Reviewer 2: ```suggestion
gems, are required to build, develop, and test this software.
```

Please make sure you have [bundler

> installed](http://bundler.io/#getting-started) on your system, and then use it

Reviewer 2: ```suggestion
installed](http://bundler.io/#getting-started) on your system, and use it
```

to install all dependencies needed for this project in the project root by

> running

Reviewer 2: ```suggestion
running:
```

```shell
$ bundle install --path .bundle/gems
```

> **NOTE:** some systems may require you to run this command with sudo.

> If you already have those gems installed, make sure they are up-to-date:

Reviewer 2: ```suggestion
If you already have those gems installed, make sure they are up to date:
```
Nerdy grammar rule: If a compound adjective doesn't come immediately before the noun, it usually doesn't require hyphenation.

```shell
$ bundle update
```

### Running tests

> With all dependencies in place and up-to-date, run the tests:

Reviewer 2: ```suggestion
With all dependencies in place and up to date, run the tests:
```

```shell
$ bundle exec rake spec
```

> This executes all the [rspec tests](http://rspec-puppet.com/) in the directories

Reviewer 2: ```suggestion
This executes all the [RSpec tests](http://rspec-puppet.com/) in the directories
```

defined
[here](https://github.com/puppetlabs/puppetlabs_spec_helper/blob/699d9fbca1d2489bff1736bb254bb7b7edb32c74/lib/puppetlabs_spec_helper/rake_tasks.rb#L17)

> and so on. rspec tests may have the same kind of dependencies as the module they

Reviewer 2: ```suggestion
and so on. RSpec tests may have the same kind of dependencies as the module they
```

are testing. Although the module defines these dependencies in its

> [metadata.json](./metadata.json), rspec tests define them in

Reviewer 2: ```suggestion
[metadata.json](./metadata.json), RSpec tests define them in
```

[.fixtures.yml](./fixtures.yml).

## Submission

### Pre-requisites

> - Make sure you have a [GitHub account](https://github.com/join)

Reviewer 2: I think we should move this one up to the first list under "## Getting started" 
```suggestion
- Make sure you have a [GitHub account](https://github.com/join).
```

> - [Open an
>   issue](https://github.com/puppetlabs/puppetlabs-http_request/issues/new/choose)
>   or [track an
>   issue](https://github.com/puppetlabs/puppetlabs-http_request/issues) you are
>   patching for

Reviewer 2: ```suggestion
- [Open an
  issue](https://github.com/puppetlabs/puppetlabs-http_request/issues/new/choose)
  or [track an
  issue](https://github.com/puppetlabs/puppetlabs-http_request/issues) you are
  patching.
```

> ### Push and PR

Reviewer 2: ```suggestion
### Push and pull request
```

> - Push your changes to your fork
> - [Open a Pull
>   Request](https://github.com/puppetlabs/puppetlabs-http_request/compare)
>   against the repository

Reviewer 2: ```suggestion
- Push your changes to your fork.
- [Open a Pull
  Request](https://github.com/puppetlabs/puppetlabs-http_request/compare)
  against the repository.
```

## More about commits 

### Make separate commits for logically separate changes.

Please break your commits down into logically consistent units which include new

> or changed tests relevant to the rest of the change.  The goal of doing this is

Reviewer 2: ```suggestion
or changed tests relevant to the rest of the change. The goal of doing this is
```

> to make the diff easier to read for whoever is reviewing your code.  In general,

Reviewer 2: ```suggestion
to make the diff easier to read for whoever is reviewing your code. In general,
```

the easier your diff is to read, the more likely someone will be happy to review
it and get it into the code base.

If you are going to refactor a piece of code, please do so as a separate commit
from your feature or bug fix changes.

> We also really appreciate changes that include tests to make sure the bug is not
> re-introduced, and that the feature is not accidentally broken.

Reviewer 2: ```suggestion
We also really appreciate changes that include tests to make sure the bug is not
re-introduced, and that the feature is not accidentally broken in a future change.
```

> Describe the technical detail of the change(s).  If your description starts to

Reviewer 2: ```suggestion
Describe the technical detail of the change(s). If your description starts to
```

get too long, that is a good sign that you probably need to split up your commit
into more finely grained pieces.

> Commits which plainly describe the things which help reviewers check the patch
> and future developers understand the code are much more likely to be merged in
> with a minimum of bike-shedding or requested changes.  Ideally, the commit
> message would include information, and be in a form suitable for inclusion in
> the release notes for the version of Puppet that includes them.

Reviewer 2: ```suggestion
A commit is much more likely to be merged with a minimum of 
bike-shedding or requested changes if you:
- Plainly describe the feature or patch that you're introducing
  with the intention of helping reviewers and future developers understand
  the code.
- Include information that will help reviewers to check and test your code.
- Include information in your commit message that would be suitable for 
  inclusion in the release notes for the version of Puppet that includes the 
  change.
```

Please also check that you are not introducing any trailing whitespace or other

> "whitespace errors".  You can do this by running `git diff --check` on your

Reviewer 2: ```suggestion
"whitespace errors". You can do this by running `git diff --check` on your
```

changes before you commit.

### Sending your patches

To submit your changes via a GitHub pull request, we _highly_ recommend that you

> have them on a topic branch, instead of directly on "main". It makes things much

Reviewer 2: ```suggestion
have them on a topic branch, instead of directly on the `main` branch. It makes things much
```

easier to keep track of, especially if you decide to work on another thing
before your first change is merged in.

GitHub has some pretty good [general documentation](http://help.github.com/) on

> using their site.  They also have documentation on [creating pull

Reviewer 2: ```suggestion
using their site. They also have documentation on [creating pull
```

requests](https://help.github.com/articles/creating-a-pull-request-from-a-fork/).

In general, after pushing your topic branch up to your repository on GitHub, you
can switch to the branch in the GitHub UI and click "Pull Request" towards the
top of the page in order to open a pull request.

### Update the related GitHub issue.

If there is a GitHub issue associated with the change you submitted, link the
issue to your pull request.

## If you have commit access to the repository

Even if you have commit access to the repository, you still need to go through

> the process above, and have someone else review and merge in your changes.  The

Reviewer 2: ```suggestion
the process above, and have someone else review and merge in your changes. The
```

# Contributing to Puppet modules

So you want to contribute to a Puppet module: Great! Below are some instructions
to get you started doing that very thing while setting expectations around code
quality as well as a few tips for making the process as easy as possible. 

## Get help

### On the web

* [General GitHub documentation](http://help.github.com/)
* [GitHub pull request
  documentation](http://help.github.com/send-pull-requests/)

### On chat

> * [Slack](https://slack.puppet.com) in the **#bolt** channel

Reviewer: I'd maybe make this it's own section right here:
```
## Community Slack Channels

Join the `#bolt` channel in the [Puppet community
Slack](https://slack.puppet.com/) where Bolt developers and community members
who use and contribute to Bolt discuss the tool.
```
If this had more than just tasks and plans in it I'd say we should link to Puppet channels too, but it's really Bolt specific.

Then have this where it was at the bottom:
```
### Github resources

* [General GitHub documentation](http://help.github.com/)
* [GitHub pull request
  documentation](http://help.github.com/send-pull-requests/)
```

## Getting started

- Fork the module repository on GitHub and clone to your workspace

> 

Reviewer: ```suggestion
```

- Make your changes!

## Commit checklist

### The basics

- My commit is a single logical unit of work
- I have checked for unnecessary whitespace with `git diff --check`
- My commit does not include commented out code or unneeded files

### The content

- My commit includes tests for the bug I fixed or feature I added
- My commit includes appropriate documentation changes if it is introducing
  a new feature or changing existing functionality
- My code passes existing test suites

### The commit message

- The first line of my commit message includes:
  - An issue number (if applicable), e.g. "(GH-xxxx) This is the first line"
  - A short description (50 characters is the soft limit, excluding ticket
    number(s))
- The body of my commit message:
  - Is meaningful
  - Uses the imperative, present tense: "change", not "changed" or "changes"
  - Includes motivation for the change, and contrasts its implementation
    with the previous behavior

## Submission

### Pre-requisites

- Make sure you have a [GitHub account](https://github.com/join)

> 

Reviewer: ```suggestion
```

- [Open an
  issue](https://github.com/puppetlabs/puppetlabs-http_request/issues/new/choose)
  or [track an
  issue](https://github.com/puppetlabs/puppetlabs-http_request/issues) you are
  patching for

### Push and PR

- Push your changes to your fork

> 

Reviewer: ```suggestion
```

the release notes for the version of Puppet that includes them.

Please also check that you are not introducing any trailing whitespace or other
"whitespace errors".  You can do this by running `git diff --check` on your
changes before you commit.

### Sending your patches

To submit your changes via a GitHub pull request, we _highly_ recommend that you
have them on a topic branch, instead of directly on "main". It makes things much
easier to keep track of, especially if you decide to work on another thing
before your first change is merged in.

GitHub has some pretty good [general documentation](http://help.github.com/) on
using their site.  They also have documentation on [creating pull
requests](https://help.github.com/articles/creating-a-pull-request-from-a-fork/).

In general, after pushing your topic branch up to your repository on GitHub, you
can switch to the branch in the GitHub UI and click "Pull Request" towards the
top of the page in order to open a pull request.

### Update the related GitHub issue.

If there is a GitHub issue associated with the change you submitted, link the
issue to your pull request.

> ## Testing

Reviewer: I'd also move this to be between 'Commit Checklist' and 'Submission'. This is honestly probably what most people are looking for - they know how to use Git, but just need the commands to run the tests.

# Contributing to Puppet modules

So you want to contribute to a Puppet module: Great! Below are some instructions
to get you started doing that very thing while setting expectations around code
quality as well as a few tips for making the process as easy as possible. 

## Getting started

- Fork the module repository on GitHub and clone to your workspace

- Make your changes!

## Commit checklist

### The basics

- [x] My commit is a single logical unit of work

> 

Reviewer: I know this is copy-pasted but I'm not sure we need all these newlines - it'd be nice to just have straight lists.

```shell
$ bundle update
```

## Running tests

With all dependencies in place and up-to-date, run the tests:

```shell
$ bundle exec rake spec
```

This executes all the [rspec tests](http://rspec-puppet.com/) in the directories
defined
[here](https://github.com/puppetlabs/puppetlabs_spec_helper/blob/699d9fbca1d2489bff1736bb254bb7b7edb32c74/lib/puppetlabs_spec_helper/rake_tasks.rb#L17)
and so on. rspec tests may have the same kind of dependencies as the module they
are testing. Although the module defines these dependencies in its
[metadata.json](./metadata.json), rspec tests define them in
[.fixtures.yml](./fixtures.yml).

# If you have commit access to the repository

Even if you have commit access to the repository, you still need to go through
the process above, and have someone else review and merge in your changes.  The
rule is that _all changes must be reviewed by a project developer that did not
write the code to ensure that all changes go through a code review process._

The record of someone performing the merge is the record that they performed the
code review. Again, this should be someone other than the author of the topic
branch.

> # Get help

Reviewer: I think, like our contributing doc, this section should be at the top.
